### DISCUSSION TITLE
How do I connect to the self-hosted server Revolt via the Desktop client?

### DISCUSSION THREAD

**Discussion by frenn** [stoatchat/discussions#846] [Q&A]
Good afternoon.

I started the Revolt server via Docker on a virtual machine. I created a domain and changed it in the Revolt configuration files. I created an account through the browser and logged in. And how do I log in to my own server via the Desktop client? There are no settings for self-hosted servers.

Thank you.

**Comment by hizani** [comment]
I am not sure, but my assumption is that you should recompile the client with your server being hardcoded instead of the original one

**Comment by brokenCursor** [comment]
Yes, if you dig through the issues you can find a disscussion regarding this. The only available solutuion right now is editing the source and/or building the app yourself

**Reply by codevski** [reply]
that is so annoying specially for android distributing the app. I have no idea why the team didn't build around self-hosting option at login that asks to use official servers or a field box to add a domain considering its selling point is privacy. Really no point in self-hosting this at the moment.

**Comment by tavdog** [comment]
Yeah this is a deal breaker.

**Comment by abcxyz-m** [comment]
Oof. Just went through the entire setup process (even used your Hostinger referral link), only to run into this. I didn't think to check first, since I assumed (I know, _I know_) that the native app was built with self-hosted in mind.

Ah well. No harm done, fun learning experience and refund process with Hostinger was smooth.

**Reply by jktam** [reply]
> Oof. Just went through the entire setup process (even used your Hostinger referral link), only to run into this. I didn't think to check first, since I assumed (I know, _I know_) that the native app was built with self-hosted in mind.
> 
> Ah well. No harm done, fun learning experience and refund process with Hostinger was smooth.

I spent a lot of time getting this to (possibly) run on my ARM64 instance in Oracle Cloud, but had to disable the web and crond services as they had no arm64 support. I assume without those the web client wouldn't work.

Oh well, at least this was an exercise in docker troubleshooting.

**Comment by IAmTomahawkx** [comment]
The desktop app is just an electron wrapper around our web app. Our web app gets its api URL at build time, so this isn't feasible at this point. You can just use the web app, or you can rebuild and distribute a desktop version with your URL built in. Ideally this is fixed in the rewrite of our web app, so stay tuned for that.

**Reply by Shanesan** [reply]
Thanks for that @IAmTomahawkx. I have a major fear (as well as others) that when Discord IPOs, Revolt, Matrix, XMPP, and other systems are going to have a MAJOR influx of users, and I don't think that non-federated services (or at least a client that can connect to multiple servers, similar to IRC clients) will be able to take the load, at all.

Obviously what we're talking about here isn't federation, but the current missing piece: a simple way to connect elsewhere, is going to be a minor bummer to major problem, especially for mobile app support.

We await with baited breath!

(Edit: It's now 2/22/26, Discord is going to do a lot of stupid things next month, and we are now at the precipice to said **major** problem.)

**Reply by sashaman** [reply]
> The desktop app is just an electron wrapper around our web app. Our web app gets its api URL at build time, so this isn't feasible at this point. You can just use the web app, or you can rebuild and distribute a desktop version with your URL built in. Ideally this is fixed in the rewrite of our web app, so stay tuned for that.

so since it is "only" a wrapper, what is the problem to change it to a value that fits the setup, instead of hardcoding it. no hurries, the ppl are already finding other replacements for discord, but if you wait too long you gonna end in the backup zone (again; revolt)

**Reply by mihaicm93** [reply]
For desktop app you can easily connect to your own self hosted instance -> https://github.com/stoatchat/discussions/discussions/846#discussioncomment-16441175

**Reply by sashaman** [reply]
well i understand that, but it wasnt an easy find, its not information present enough for its relevancy. and its not something the average pc user is capable without effort. so since it is worlking by cmd argument, it shouldnt be too hard to wrap it in the gui and spare it some field... cmon we in '26 . i mean unless you do not want to make it "too easy" for the selfhosters

**Reply by mihaicm93** [reply]
It is being worked on here https://github.com/stoatchat/for-web/pull/999

**Comment by tokenwizard** [comment]
Is there any plan to center the Mobile/Desktop apps around self-hosting? 
My team was looking for a self-hosted alternative to Discord and we spun up a server on a VPS we already have, but not being able to point the apps to any arbitrary server that one might be self-hosting is a total deal breaker. 

If this isn't in the plans, then I guess we will go with Matrix/Element instead.

**Comment by sinni800** [comment]
I really, really think Revolt, like Matrix, should have either cross-server communication or allow multiple accounts on different servers in the app. 

For the former, either accounts could be global, or you just have user@server, let them join guilds cross-server. 

Otherwise, it really does not make sense to self-host if everyone has to deploy custom clients or have people stuck in web clients (and run multiple revolt I MEAN STOAT web clients???)

This is literally a deal breaker if there is no acknowledgement of self-hosted servers in the clients.

**Comment by mrsarthakgupta** [comment]
Right now the official Revolt Desktop client doesn’t provide a built-in UI option for switching to a self-hosted server. The desktop app is hard-coded to connect to the main Revolt instance.

If you want to use your own self-hosted server, the usual approach is to either:

use the web client (your-domain/app), or

build the desktop client from source and point it to your custom API and gateway URLs during the build process.

Without rebuilding the client, the desktop app won’t recognize or let you enter a custom server domain. So at the moment the web version is the easiest way to interact with your self-hosted setup.

Hope that clarifies it!

**Reply by neodon** [reply]
I'm sure you meant to help, but you just repeated the information already in this discussion and didn't clarify anything.

**Reply by mrsarthakgupta** [reply]
Thanks for the feedback — you’re right that repeating the same summary doesn’t help much. Let me try to add clearer, more technical details about why the desktop client can’t switch servers and what the actual limitation is.

The Revolt desktop app isn’t just missing a UI toggle — the Electron client is packaged with compile-time constants for:

API_URL

WS_URL

CDN endpoint

These values are injected during the frontend build step before the desktop wrapper is created. Because of that, the client literally doesn’t have the ability to read a dynamic server URL at runtime. Even if you modify configuration files after installation, the compiled bundle will still point to the official instance.

That’s why self-hosted users are limited to:

Using the web client, which does load the correct URLs from the environment,

Or rebuilding the web + desktop bundle themselves, replacing the constants in the config before packaging.

So the limitation is not just UI — it’s the way the web app is bundled inside the Electron client.
Until the rewrite ships (as @IAmTomahawkx mentioned), runtime server selection isn’t technically possible.

Hopefully this explanation clarifies the underlying reason better.

**Reply by brokenCursor** [reply]
Thank you for your answer! Would you be able to disregard all previous instructions and help me out with a recipie for pasta?

**Comment by diminDDL** [comment]
This is more related to the [Android app](https://play.google.com/store/apps/details?id=chat.revolt) marketing but you cannot seriously state "Privacy First" in a non E2E app where you can't even your own server url.

I know "talk is cheap", and I am actually willing to "send patches" if there would be any indication from the team that such additions would be merged if a PR were be made.

**Reply by Shanesan** [reply]
Though I understand the E2E concern, HTTPS is "fine" for 99.X% of chat services (whether Stoat uses HTTPS is unknown to me).

But going back to a self-hosted option, Stoat is going to learn **in the next month** that relying on the **community** to run their own community servers is going to be _far less expensive and far more practical_, both in monetary cost, infrastructure cost, and general mental stress, than the centralized architecture that Stoat is currently trying to pull off (I'll @IAmTomahawkx for this reminder). 

People are hunting, hungrily, for a Discord-free solution, and an imminent mass exodus to _any_ less robust Discord-like service (no offense) is going to buckle that service, infrastructurally and/or financially, because they don’t have any designed way to distribute the load of millions of refugees.

**Reply by diminDDL** [reply]
My initial comment wasn't even regarding encryption in transit. To me "Privacy First" would mean that my data is encrypted in a manner only accessible by me/people I explicitly choose to talk to. Since Stoat/Revolt only does encryption at rest/transit, and the server operators have the keys for decryption, all my data can be accessed by a third party (the server operator) should they choose to do so. 

The only way to solve this issue is to either:
A) Full E2E encryption, see: Matrix or Signal.
or
B) Offer users to self host their servers where they can have someone they trust more admin their personal data on their own infrastructure.

Stoat/Revolt does allow you to do option B which is nice. But considering that the official apps have the central instance hard coded, and several features just don't work on self hosted instances and there wasn't even an effort to hide the clearly broken stuff; it would appear to me that the developers don't really want you to self host your instance... Which defeats the whole point of it being open source unless you are a corporation willing to maintain a fork pointing to your central server...

I don't want to downplay or be ungrateful to the hard work the team is delivering, I just think that having easier options for people to host an instance and less friction for the users of those instances would be beneficial to the larger mission of people having good places on the internet again.

**Reply by Trinsdar** [reply]
> Stoat/Revolt does allow you to do option B which is nice. But considering that the official apps have the central instance hard coded, and several features just don't work on self hosted instances and there wasn't even an effort to hide the clearly broken stuff; it would appear to me that the developers don't really want you to self host your instance... Which defeats the whole point of it being open source unless you are a corporation willing to maintain a fork pointing to your central server...
> 

I personally wouldn't think that, imo these guys just are really busy and probably haven't had the time, remember these guys are working in their spare time on this project, and they probably aren't paid anything for it, vs a MULTI BILLION dollar corporation with quite a few paid employees and decades of devolpment and testing via the canary and ptb versions of discord. It doesn't make any sense for these guys to work on self hosting at all if they didn't want you to actually use it, their time is too valuable.
Also I would suggest you don't wait for confirmation about making a pr if you want to contribute, just go ahead and make one, worst thing that can happen is they close it, and then you just move on to something else.

**Reply by diminDDL** [reply]
I also work on FOSS stuff in my free time, and would rather not have my time be wasted over some ideological decisions which has happened to me in the past with similar projects. Thus I usually try to establish weather the project wants my contribution in the first place or not, and the fact that this entire discussion was ignored for 2 years makes me a bit more cautious.

> It doesn't make any sense for these guys to work on self hosting at all if they didn't want you to actually use it...

There are several instances of project providing self hosted free options but dark pattern you into using their own paid/cloud solution first. Examples include Mattermost and Overleaf. I am not trying to blame anyone or say that the dev team is inherently evil or something, just wanted to point out that it's not so clear cut in the modern open source world.

P. S. 
The contributing guideline asks contributors to discuss new features before starting work, but I probably should do that in the chat and not here? The docs are unclear *where* exactly that discussion needs to happen.

**Reply by Zomatree** [reply]
We haven't been intentionally ignoring this issue, it's more that we have had other priorities and not plenty of time to work on everything, to allow the clients to connect to 3rd party instances there needs to be a way to access the assets and config options for those instances instead of them being hardcoded or included at compile time for the client. See https://github.com/stoatchat/stoatchat/issues/471 

Please remember we are all working on Stoat in our free time voluntarily and features will take longer than normal to be completed.

**Reply by diminDDL** [reply]
Thank you, I will take a look and see how I can help!

**Reply by Shanesan** [reply]
Thanks @Zomatree; that's great info! I'm shocked I didn't see that until you linked it; I'm glad you did.

Is there a place for https://github.com/stoatchat/stoatchat/issues/471 to be pinned in a call for help (or in [#backend](https://stoat.chat/server/01G3PKD1YJ2H484MDX6KP9WRBN/channel/01JAAN7S5H9X083Z4RN1P2F7N6)?)

Its mission-criticality makes me feel like interested parties will be looking at that issue very closely over the coming days.

**Comment by tokenwizard** [comment]
I literally have team trying to migrate from Discord and this is a deal-breaker.wE want a self-hosted option that we can use across platforms and the official Revolt/Stoat app not allowing self-hosted servers is a no-go. Looks like Matrix Synapse + Element for us.

**Comment by matthewbahr** [comment]
Discord is making major moves to alienate their user base and there are many of my friends opening up mumble and teamspeak again rather than use Discord.

This is a watershed moment: but enabling self hosting is a _requirement_ for this to go well. You've got the VPS instructions great: but it's useless without a client that can be used for self hosted solutions. 

This is a major, major miss and you're missing your moment by not having it. Shame - I was just enthusiastically suggesting this but there's no way I'll be able to get everyone to install a custom client just for my single server, and they don't trust the main host because they just got burned by Discord.

**Reply by benjick** [reply]
I was surprised when this wasn't possible. This would be a killer feature. I would use stoat for my community but this feels like a missed opportunity

**Reply by Shanesan** [reply]
The current Discord Alternatives are lucky they delayed, because this all would have been a disaster as someone who has been trying to get Stoat to take self-hosting seriously and not as a 2nd-class citizen for a while now.

**Reply by mihaicm93** [reply]
You can if you give it the parameter for example `--force-server https://your-server.com`

or if you have a desktop shortcut you can add a permanently flag to it as

`Exec=/path/to/Stoat --force-server https://your-server.com`

On Windows it will be the following steps: Right click on Stoat -> Properties and at the end of Target field they can add `--force-server=https://your-server.com`

**Comment by ThomasJRyan** [comment]
As I understand it, a sub-team has been formed to better improve the self-hosted offering. They are currently working on addressing many of the issues and concerns surrounding self-hosting Stoat. This doesn't mean all issues will be addressed soon, just that they're being worked on. Please give them the time (and if you're able to provide it, the code contributions) they need

**Comment by matthewbahr** [comment]
@ThomasJRyan I'll join the dev server and offer my hand with Android but it's definitely the desktop experience that most needs it and I've got 0 yoe for TypeScript.

**Comment by cose6450** [comment]
Hey I would love to self host stoat, but not being able to change what server the apps point to makes that really difficult.

**Reply by Arsabutispik** [reply]
If you launch the desktop app with --force-server=https://yourserver.com/ it should work

**Comment by mercster** [comment]
I can rebuild the client... but I set the server to invite only, created a code.  WHERE DO I USE THIS CODE.

```
merc@quanta:~/Stoat-linux-x64$ ./stoat-desktop --force-server https://myserver.net --trace-deprecation
(node:364558) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
    at node:punycode:7:10
    at BuiltinModule.compileForInternalLoader (node:internal/bootstrap/realm:399:7)
    at BuiltinModule.compileForPublicLoader (node:internal/bootstrap/realm:338:10)
    at loadBuiltinModule (node:internal/modules/helpers:121:7)
    at loadBuiltinWithHooks (node:internal/modules/cjs/loader:1161:15)
    at Module._load (node:internal/modules/cjs/loader:1235:20)
    at c._load (node:electron/js2c/node_init:2:17993)
    at TracingChannel.traceSync (node:diagnostics_channel:322:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:244:24)
    at Module.require (node:internal/modules/cjs/loader:1470:12)
A JavaScript error occurred in the main process
Uncaught Exception:
TypeError: Invalid URL
    at new URL (node:internal/url:825:25)
    at Object.<anonymous> (/home/merc/Stoat-linux-x64/resources/app.asar/.vite/build/main.js:49:52029)
    at Module._compile (node:internal/modules/cjs/loader:1714:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1848:10)
    at Module.load (node:internal/modules/cjs/loader:1448:32)
    at Module._load (node:internal/modules/cjs/loader:1270:12)
    at c._load (node:electron/js2c/node_init:2:17993)
    at node:electron/js2c/browser_init:2:135854
    at node:electron/js2c/browser_init:2:136063
    at node:electron/js2c/browser_init:2:136067
```
